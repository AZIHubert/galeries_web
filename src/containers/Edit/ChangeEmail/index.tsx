import * as React from 'react';
import { useFormik } from 'formik';

import { changeEmailSchema } from '#helpers/schemas';

const initialValues = {
  password: '',
};

interface ChangeEmailI {
  loading: boolean;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
}

const ChangeEmail = ({ loading, setLoading }: ChangeEmailI) => {
  const formik = useFormik({
    initialValues,
    onSubmit: () => {
      if (!loading) { setLoading(true); }
    },
    validateOnChange: false,
    validateOnBlur: true,
    validationSchema: changeEmailSchema,
  });
  return (
    <div>
      <p>
        change your email
      </p>
      <p>
        Enter your password to receive a message on your current adress.
        This message contain a link to update your email.
      </p>
      <form onSubmit={formik.handleSubmit}>
        <label
          htmlFor="password"
        >
          password
        </label>
        <input
          data-testid='passwordField'
          disabled={loading}
          id='password'
          name='password'
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          type='password'
          value={formik.values.password}
        />
        {formik.errors.password
          && formik.touched.password && (
          <div
            data-testid='passwordError'
          >
            {formik.errors.password}
          </div>
        )}
        <button
          disabled={loading}
          data-testid='submitButton'
          type='submit'
        >
          {loading ? 'loading' : 'change your email'}
        </button>
      </form>
    </div>
  );
};

export default ChangeEmail;
