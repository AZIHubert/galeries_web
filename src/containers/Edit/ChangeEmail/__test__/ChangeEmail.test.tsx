import * as React from 'react';
import renderer from 'react-test-renderer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import {
  REQUIRED,
} from '#helpers/formErrors';

import ChangeEmail from '../index';

const Container = () => {
  const [loading, setLoading] = React.useState(false);
  return (
    <ChangeEmail
      loading={loading}
      setLoading={setLoading}
    />
  );
};

describe('ChangeEmail', () => {
  const setLoading = jest.fn;
  let passwordField: HTMLElement;
  let submitButton: HTMLElement;
  beforeEach(() => {
    const { getByTestId } = render(<Container />);
    passwordField = getByTestId('passwordField');
    submitButton = getByTestId('submitButton');
  });
  afterEach(cleanup);
  it('renders without crashing', () => {
    const tree = renderer.create(
      <ChangeEmail
        loading={false}
        setLoading={setLoading}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should not loading if errors', async () => {
    fireEvent.click(submitButton);
    const passwordError = await screen.findByTestId('passwordError');
    expect(passwordField).not.toBeDisabled();
    expect(passwordError).toHaveTextContent(REQUIRED);
    expect(submitButton).not.toBeDisabled();
    expect(submitButton).toHaveTextContent('change your email');
  });
  it('should loading if no error', async () => {
    fireEvent.change(passwordField, { target: { value: 'password' } });
    fireEvent.click(submitButton);
    await screen.findByText('loading');
    expect(passwordField).toBeDisabled();
    expect(submitButton).toBeDisabled();
  });
  describe('should return an error if', () => {
    describe('password', () => {
      it('is empty', async () => {
        fireEvent.blur(passwordField);
        const passwordError = await screen.findByTestId('passwordError');
        expect(passwordError).not.toBeNull();
        expect(passwordError).toHaveTextContent(REQUIRED);
      });
    });
  });
});
