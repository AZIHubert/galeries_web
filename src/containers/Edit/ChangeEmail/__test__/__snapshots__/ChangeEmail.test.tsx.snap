// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeEmail renders without crashing 1`] = `
<div>
  <p>
    change your email
  </p>
  <p>
    Enter your password to receive a message on your current adress. This message contain a link to update your email.
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="password"
    >
      password
    </label>
    <input
      data-testid="passwordField"
      disabled={false}
      id="password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      type="password"
      value=""
    />
    <button
      data-testid="submitButton"
      disabled={false}
      type="submit"
    >
      change your email
    </button>
  </form>
</div>
`;
